File: lab1/Shape.cpp
#include "Shape.h"
#include <chrono>

static int ids = 0;

Shape::Shape(int ix, int iy) {
	timestamp = std::chrono::system_clock::now().time_since_epoch().count();
	id = ids++;
	x = ix;
	y = iy;
}

int Shape::left() {
	return x;
}

int Shape::right() {
	return x;
}
int Shape::top() {
	return x;
}

int Shape::bottom() {
	return x;
}

int Shape::compId(Shape* s)
{
	return id == s->id ? 0 : id > s->id ? 1 : -1;
}

int Shape::compTime(Shape* s) {
	return timestamp == s->timestamp ? 0 : timestamp > s->timestamp ? 1 : -1;
}

Shape::Shape() : Shape(0,0) {}

File: lab3/src/main/java/com/example/models/Shape.java
package com.example.models;

import jakarta.persistence.*;

import java.time.Instant;

import com.example.annotations.*;

@Entity
@Inheritance(strategy = InheritanceType.JOINED)
public abstract class Shape {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Get Integer id;

    private @Get long timestamp = Instant.now().getEpochSecond();

    @Get @Set int x;
    @Get @Set int y;

    @ManyToOne
    @JsonExclude
    private ShapesGroup groupEntity;
    
    public String groupName() { 
    	return groupEntity == null ? "<no group>" : groupEntity.name; 
    }
    
    public void setGroup(@Named("groupId") ShapesGroup g) { 
    	System.out.println("GROUP SET: " + g);
    	this.groupEntity = g; 
    }

    protected Shape() {}

    protected Shape(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public abstract int left();
    public abstract int right();
    public abstract int top();
    public abstract int bottom();

    public long lifetime() {
        return Instant.now().getEpochSecond() - timestamp;
    }

    public void refresh() {
        this.timestamp = Instant.now().getEpochSecond();
    }

}
